accord-core/src/main/java/accord/local/CommandStores.java

```diff
@@ -181,17 +179,59 @@ public long shards(AbstractKeys<?, ?> keys, long minEpoch, long maxEpoch)
             return accumulate;
         }
 
-        KeyRanges currentRanges()
+        public <T extends Sortable> long shards(Sortables<T, ?> keysOrRanges, long minEpoch, long maxEpoch)
+        {
+            long terminalValue = -1L >>> (32 - shards.length);
+            switch (keysOrRanges.kindOfContents())
+            {
+                default: throw new AssertionError();
+                case Key:
+                {
+                    long accumulate = 0L;
+                    for (int i = Math.max(0, indexForEpoch(minEpoch)), maxi = indexForEpoch(maxEpoch); i <= maxi ; ++i)
+                    {
+                        accumulate = Sortables.foldl((AbstractKeys<?, ?>)keysOrRanges, ranges[i], ShardedRanges::addKeyIndex, shards.length, accumulate, terminalValue);
+                    }
+                    return accumulate;
+                }
+
+                case Range:
+                {
+                    long accumulate = 0L;
+                    for (int i = Math.max(0, indexForEpoch(minEpoch)), maxi = indexForEpoch(maxEpoch); i <= maxi ; ++i)
+                    {
+                        // include every shard if we match a range
+                        accumulate = Sortables.foldl((Ranges)keysOrRanges, ranges[i], (idx, k, p, a) -> p, terminalValue, accumulate, terminalValue);
+                    }
+                    return accumulate;
+                }
+            }
+        }
+
+        long shard(RoutingKey scope, long minEpoch, long maxEpoch)
+        {
+            long result = 0L;
+            for (int i = Math.max(0, indexForEpoch(minEpoch)), maxi = indexForEpoch(maxEpoch); i <= maxi ; ++i)
+            {
+                int index = ranges[i].indexOf(scope);
+                if (index < 0)
+                    continue;
+                result = addKeyIndex(0, scope, shards.length, result);
+            }
+            return result;
+        }
+
+        Ranges currentRanges()
         {
             return ranges[ranges.length - 1];
         }
 
-        static long keyIndex(RoutingKey key, long numShards)
+        static long keyIndex(SortableKey key, long numShards)
```

Comment:
`RoutableKey`?

Reply:
As discussed offline, `CommandStores` is happy to accept either kind of key